### PR TITLE
Fixes #30719 - allow spawning without shell

### PR DESCRIPTION
**Pull request by lzap**
For https://github.com/theforeman/smart_proxy_shellhooks we need to be able to spawn a command through the utility helper without shell to prevent possible problems. Popen3 supports this via multiple arguments, but the utility helper did not allow passing an array.

Also logging was improved ("Started task" now contains also the thread PID) and logging of plain/text body was improved for easy to understand examples in the plugin README.

### CODE REVIEW

File: lib/proxy/util.rb
Comment (ekohl) [inline lib/proxy/util.rb]: I thought `aninput` as hard to read. Is `input` a reserved value/method? If so, I think `input_value` might be easier to read.

File: lib/proxy/util.rb
Comment (ekohl) [inline lib/proxy/util.rb:20]: This still looks useful.

**Review by ekohl — COMMENTED** [review]
Small details, overall this makes sense. I'm certainly a big fan of passing arrays as commands rather than trying to shell escape.

Does make me wonder: should we update HttpDownload as well to use this? After Puppetrun is removed, it's the only place that still uses `escape_for_shell`. I'm fine with doing that in a follow up PR.

File: lib/proxy/util.rb
Comment (ares) [reply lib/proxy/util.rb]: I wish we also renamed `acommand` to just `command` at the same time. I know it's not changed by this PR :-)

Comment (lzap) [reply lib/proxy/util.rb]: Sure it's rare naming convention and I just followed what was there but can refactor.

File: lib/proxy/util.rb
Comment (lzap) [reply lib/proxy/util.rb:20]: Absolutely, I just moved it few lines down so I could actually use thread ID in the message so it's obvious which thread ID messages belongs to which command.

**Comment by lzap** [comment]
No problem, I can incorporate the wget change into this PR. Amended, tested. Looks way cleaner now, no escaping anymore.

File: lib/proxy/util.rb
Comment (ekohl) [reply lib/proxy/util.rb:20]: The thing is that you now move it to within the `Thread.new` and `popen3`. If either fails, you don't really know. For example, `popen3` can raise an exception. While the command may not say a huge amount, I'm not sure what the error output is if the command in the thread failed. Those were my considerations for keeping it in. Since it's debug, most users won't see it, but can be useful when things break for a weird reason.

Not going to block on it because it may be a bit of an edge case so I'll leave it up to you.

File: lib/proxy/util.rb
Comment (ekohl) [inline lib/proxy/util.rb:25]: Coming from Python, it surprised me you can do this with strings. In Python `my-cmd` would end up as `['m', 'y', '-', 'c', 'm', 'd']`. I learned that here Python and Ruby differ.

**Review by ekohl — APPROVED** [review]

File: lib/proxy/util.rb
Comment (lzap) [reply lib/proxy/util.rb:25]: Yeah I tested this. There's a test for this as well.

Comment (ekohl) [reply lib/proxy/util.rb:25]: I did indeed test it in a shell. Just one of those Today I Learned moments.

**Comment by ekohl** [comment]
I'm going to assume @ares has no objections other than the one he mentioned. Since that's addressed, I'm going to merge this. Thanks!

After merging this and the puppetrun removal I want to deprecate the shell escape method in a separate PR.

**Pull request merged**
